File: week17/sangmin/problem5/Problem5.java
package week17.sangmin.problem5;

import java.util.*;

class Solution {

    String str = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    Map<String, Integer> map = new HashMap<>();

    void init(){
        for(int idx = 0; idx < str.length(); idx++){
            map.put(str.charAt(idx) + "", idx + 1);
        }
    }

    public int[] solution(String msg) {
        List<Integer> list = new ArrayList<>();
        init();
        for(int idx = 0; idx < msg.length(); idx++){
            boolean isEnd = true;
            StringBuilder sb = new StringBuilder();
            sb.append(msg.charAt(idx));
            for(int nextIdx = idx + 1; nextIdx < msg.length(); nextIdx++) {
                if (!map.containsKey(sb.append(msg.charAt(nextIdx)).toString())) {
                    idx = nextIdx - 1;
                    isEnd = false;
                    break;
                }
            }
            if(!isEnd) {
                if(!map.containsKey(sb.toString())){
                    map.put(sb.toString(), map.size() + 1);
                }
                String output = sb.toString().substring(0, sb.length() - 1);
                list.add(map.get(output));
            }else{
                String output = sb.toString();
                list.add(map.get(output));
                break;
            }
        }
        int[] answer = new int[list.size()];
        for(int i = 0; i < list.size(); i++){
            answer[i] = list.get(i);
        }
        return answer;
    }
}

public class Problem5 {
    public static void main(String[] args) {
//        new Solution().solution("KAKAO");
        new Solution().solution("TOBEORNOTTOBEORTOBEORNOT");
    }
}

File: week17/juyoung/problem5/problem5.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
using namespace std;

#include <string>
#include <vector>
#include <map>
using namespace std;

vector<int> solution(string msg) {
	vector<int> answer;
	map<string, int> dictionary;
	int idx = 27;

	// 1. 길이가 1인 모든 단어를 포함하도록 사전을 초기화한다.
	string str = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
	for (int i = 0; i < 26; i++)
		dictionary[str.substr(i, 1)] = i + 1;

	for (int m = 0; m < msg.length(); m++) {
		string w = "";

		// 2. 사전에서 현재 입력과 일치하는 가장 긴 문자열 w를 찾는다.
		int i;
		for (i = m; i < msg.length(); i++) {
			w += msg[i];
			if (dictionary[w] > 0) continue;
			else {
				w = w.substr(0, w.length() - 1);
				break;
			}
		}

		// 3. w에 해당하는 사전의 색인 번호를 출력한다.
		// for문을 통해 자동으로 다음 문자를 검색하므로 입력에서 따로 w를 제거하지는 않는다.
		answer.push_back(dictionary[w]);

		// 4. 입력에서 처리되지 않은 다음 글자가 남아있다면(c), w+c에 해당하는 단어를 사전에 등록한다.
		if (m < msg.length() - 1) {
			if (dictionary[w + msg[i]] > 0);
			else dictionary[w + msg[i]] = idx++;
		}

		m = i - 1;
	}

	return answer;
}

int main() {
	string msg;
	vector<int> answer;

	cin >> msg;
	answer = solution(msg);
	
	for (int i = 0; i < (int)answer.size(); i++)
		cout << answer[i] << " ";
	return 0;
}

/*
KAKAO
TOBEORNOTTOBEORTOBEORNOT
ABABABABABABABAB
*/